### PR TITLE
[bde] Updated to 3.117.0

### LINKED ISSUE #18937: [bde] build failure when python is python3
**Host Environment**
- OS: Ubuntu 20.04
- Compiler: GCC 9

**To Reproduce**
`./vcpkg install bde`

**Failure logs** 
```
Building package bde[core]:x64-linux...
-- Using /home/bion/vcpkg/downloads/bloomberg-bde-tools-d4b1a7670829228f4ec81ecdccc598ce03ae8e80.tar.gz
-- Cleaning sources at /mnt/vcpkg-ci/buildtrees/bde/src/ce03ae8e80-ef877fdb7c.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source /home/bion/vcpkg/downloads/bloomberg-bde-tools-d4b1a7670829228f4ec81ecdccc598ce03ae8e80.tar.gz
-- Using source at /mnt/vcpkg-ci/buildtrees/bde/src/ce03ae8e80-ef877fdb7c.clean
-- Configure bde-tools-v1.x
-- Using /home/bion/vcpkg/downloads/bloomberg-bde-3720d132d0879f19b9084cca62ebc75f1f24e1a3.tar.gz
-- Cleaning sources at /mnt/vcpkg-ci/buildtrees/bde/src/5f1f24e1a3-5405076f04.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source /home/bion/vcpkg/downloads/bloomberg-bde-3720d132d0879f19b9084cca62ebc75f1f24e1a3.tar.gz
-- Using source at /mnt/vcpkg-ci/buildtrees/bde/src/5f1f24e1a3-5405076f04.clean
-- Configuring x64-linux-dbg
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:105 (message):
    Command failed: waf configure --library-type=static --use-flat-include-dir --abi-bits=64 --prefix=/mnt/vcpkg-ci/packages/bde_x64-linux/debug --out=/mnt/vcpkg-ci/buildtrees/bde/x64-linux-dbg/bde-3.2.0.0 --build-type=debug
    Working Directory: /mnt/vcpkg-ci/buildtrees/bde/src/5f1f24e1a3-5405076f04.clean
    Error code: 2
    See logs for more information:
      /mnt/vcpkg-ci/buildtrees/bde/configure-x64-linux--dbg-out.log

Call Stack (most recent call first):
  ports/bde/portfile.cmake:67 (vcpkg_execute_required_process)
  scripts/ports.cmake:139 (include)


Error: Building package bde:x64-linux failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `./vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: bde:x64-linux
  Vcpkg version: 2021-05-05-unknownhash
```

```
Traceback (most recent call last):
  File "/mnt/vcpkg-ci/buildtrees/bde/src/ce03ae8e80-ef877fdb7c.clean/bin/.waf3-1.9.9-4469eb5c219e16263f641e0916ce6a48/waflib/Node.py", line 295, in ant_iter
    raise StopIteration
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/mnt/vcpkg-ci/buildtrees/bde/src/ce03ae8e80-ef877fdb7c.clean/bin/.waf3-1.9.9-4469eb5c219e16263f641e0916ce6a48/waflib/Scripting.py", line 120, in waf_entry_point
    run_commands()
  File "/mnt/vcpkg-ci/buildtrees/bde/src/ce03ae8e80-ef877fdb7c.clean/bin/.waf3-1.9.9-4469eb5c219e16263f641e0916ce6a48/waflib/Scripting.py", line 177, in run_commands
    parse_options()
  File "/mnt/vcpkg-ci/buildtrees/bde/src/ce03ae8e80-ef877fdb7c.clean/bin/.waf3-1.9.9-4469eb5c219e16263f641e0916ce6a48/waflib/Scripting.py", line 150, in parse_options
    Context.create_context('options').execute()
  File "/mnt/vcpkg-ci/buildtrees/bde/src/ce03ae8e80-ef877fdb7c.clean/bin/.waf3-1.9.9-4469eb5c219e16263f641e0916ce6a48/waflib/Options.py", line 149, in execute
    super(OptionsContext,self).execute()
  File "/mnt/vcpkg-ci/buildtrees/bde/src/ce03ae8e80-ef877fdb7c.clean/bin/.waf3-1.9.9-4469eb5c219e16263f641e0916ce6a48/waflib/Context.py", line 88, in execute
    self.recurse([os.path.dirname(g_module.root_path)])
  File "/mnt/vcpkg-ci/buildtrees/bde/src/ce03ae8e80-ef877fdb7c.clean/bin/.waf3-1.9.9-4469eb5c219e16263f641e0916ce6a48/waflib/Context.py", line 129, in recurse
    user_function(self)
  File "/mnt/vcpkg-ci/buildtrees/bde/src/5f1f24e1a3-5405076f04.clean/wscript", line 27, in options
    ctx.load('bdebuild.waf.wscript', tooldir=_get_tools_path(ctx))
  File "/mnt/vcpkg-ci/buildtrees/bde/src/ce03ae8e80-ef877fdb7c.clean/bin/.waf3-1.9.9-4469eb5c219e16263f641e0916ce6a48/waflib/Context.py", line 85, in load
    fun(self)
  File "/mnt/vcpkg-ci/buildtrees/bde/src/ce03ae8e80-ef877fdb7c.clean/lib/python/bdebuild/waf/wscript.py", line 73, in options
    ctx.load('compiler_c')
  File "/mnt/vcpkg-ci/buildtrees/bde/src/ce03ae8e80-ef877fdb7c.clean/bin/.waf3-1.9.9-4469eb5c219e16263f641e0916ce6a48/waflib/Context.py", line 85, in load
    fun(self)
  File "/mnt/vcpkg-ci/buildtrees/bde/src/ce03ae8e80-ef877fdb7c.clean/bin/.waf3-1.9.9-4469eb5c219e16263f641e0916ce6a48/waflib/Tools/compiler_c.py", line 40, in options
    opt.load_special_tools('c_*.py',ban=['c_dumbpreproc.py'])
  File "/mnt/vcpkg-ci/buildtrees/bde/src/ce03ae8e80-ef877fdb7c.clean/bin/.waf3-1.9.9-4469eb5c219e16263f641e0916ce6a48/waflib/Context.py", line 319, in load_special_tools
    lst=self.root.find_node(waf_dir).find_node('waflib/extras').ant_glob(var)
  File "/mnt/vcpkg-ci/buildtrees/bde/src/ce03ae8e80-ef877fdb7c.clean/bin/.waf3-1.9.9-4469eb5c219e16263f641e0916ce6a48/waflib/Node.py", line 344, in ant_glob
    ret=[x for x in self.ant_iter(accept=accept,pats=[to_pat(incl),to_pat(excl)],maxdepth=kw.get('maxdepth',25),dir=dir,src=src,remove=kw.get('remove',True))]
  File "/mnt/vcpkg-ci/buildtrees/bde/src/ce03ae8e80-ef877fdb7c.clean/bin/.waf3-1.9.9-4469eb5c219e16263f641e0916ce6a48/waflib/Node.py", line 344, in <listcomp>
    ret=[x for x in self.ant_iter(accept=accept,pats=[to_pat(incl),to_pat(excl)],maxdepth=kw.get('maxdepth',25),dir=dir,src=src,remove=kw.get('remove',True))]
RuntimeError: generator raised StopIteration

```

**Additional context**
Ubuntu 20.04 contains no out of the box way to make python mean python2.

### PR DESCRIPTION
**Pull request by adamncasey**
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] SHA512s are updated for each updated download
- [x] The "supports" clause reflects platforms that may be fixed by this new version
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

Fixes https://github.com/microsoft/vcpkg/issues/18937

Replaces https://github.com/microsoft/vcpkg/pull/31644

Changes since the prior PR:

1) Remove the cxx17 feature flag. This is an incomplete solution to the issue but avoids allowing conflicting feature flags in libraries.
2) Disable OSX build. Linux is the primary target - I'd like to fix others in future version of this port. The existing BDE version in vcpkg is ancient & seemingly non-functional.

### CODE REVIEW

**Comment by adamncasey** [comment]
CC'ing @BillyONeal and @FrankXie05 from the previous PR

File: ports/bde/portfile.cmake [hunk 1]
```diff
@@ -1,133 +1,72 @@
...
 message(STATUS "Configure bde-tools-v${BDE_TOOLS_VERSION}")
 if(VCPKG_CMAKE_SYSTEM_NAME)
-    set(ENV{PATH} "$ENV{PATH}:${PYTHON2_EXE_PATH}")
+    set(ENV{PATH} "$ENV{PATH}:${PYTHON3_EXE_PATH}")
     set(ENV{PATH} "$ENV{PATH}:${TOOLS_PATH}/bin")
 else()
-    set(ENV{PATH} "$ENV{PATH};${PYTHON2_EXE_PATH}")
+    set(ENV{PATH} "$ENV{PATH};${PYTHON3_EXE_PATH}")
     set(ENV{PATH} "$ENV{PATH};${TOOLS_PATH}/bin")
 endif()
```
Comment (BillyONeal) [inline ports/bde/portfile.cmake:30]:
```suggestion
vcpkg_add_to_path("${PYTHON3_EXE_PATH}")
vcpkg_add_to_path("${TOOLS_PATH}/bin")
```

File: ports/bde/portfile.cmake [hunk 2]
```diff
@@ -1,133 +1,72 @@
 vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
-set(CONFIGURE_COMMON_ARGS ${CONFIGURE_COMMON_ARGS} --library-type=static)
 
-set(BDE_VERSION 3.2.0.0)
-set(BDE_TOOLS_VERSION 1.x)
+set(BDE_VERSION 3.117.0.0)
```
Comment (BillyONeal) [inline ports/bde/portfile.cmake:3]:
```suggestion
```

And just use the builtin `${VERSION}` instead?

File: ports/bde/portfile.cmake [hunk 3]
```diff
@@ -1,133 +1,72 @@
...
+set(BDE_TOOLS_VERSION 3.117.0.0)
```
Comment (BillyONeal) [inline ports/bde/portfile.cmake:4]:
```suggestion
set(BDE_TOOLS_VERSION "${VERSION}")
```

Unless this as a separate concept no longer need exist?

File: ports/bde/portfile.cmake [hunk 4]
```diff
@@ -1,133 +1,72 @@
...
 
 # Paths used in build
 set(SOURCE_PATH_DEBUG ${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg/bde-${BDE_VERSION})
 set(SOURCE_PATH_RELEASE ${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel/bde-${BDE_VERSION})
```
Comment (BillyONeal) [inline ports/bde/portfile.cmake:9]:
```suggestion
```

This looks dead now?

File: ports/bde/portfile.cmake [hunk 5]
```diff
@@ -1,133 +1,72 @@
...
-    SHA512 234ebb71997f5b7d3951584235ead10f977689cef323ae1c198629a6b1995b1481d8a1515d827c46df10209bdc66e1f3cc7780dafee9ca0ff4172be47c460d78
-    HEAD_REF master
+    REF 3.117.0.0
+    SHA512 810b4a06a08739dcd990751dd543aa7dc58355f9d64a7c96ef0cf45c81501946434db42ad5bcf5d16110d5a463586b587ce09a446136e824298f39a8a871b490
+    HEAD_REF 3.117.0.0
 )
 
 # Clean up previous builds
 file(REMOVE_RECURSE ${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel
                     ${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg)
```
Comment (BillyONeal) [inline ports/bde/portfile.cmake:43]:
```suggestion
```

File: ports/bde/portfile.cmake [hunk 6]
```diff
@@ -1,133 +1,72 @@
...
-    set(CONFIGURE_COMMON_ARGS ${CONFIGURE_COMMON_ARGS} --abi-bits=64)
-else()
-    message(FATAL_ERROR "Unsupported target architecture: ${VCPKG_TARGET_ARCHITECTURE}")
-endif()
-if(VCPKG_TARGET_IS_WINDOWS)
-    set(CONFIGURE_COMMON_ARGS ${CONFIGURE_COMMON_ARGS} --msvc-runtime-type=static)
+set(BDE_OPTIONS "-DBDE_BUILD_TARGET_CPP17=ON" "-DCMAKE_CXX_STANDARD=17" "-DCMAKE_CXX_STANDARD_REQUIRED=ON" "-DCMAKE_CXX_EXTENSIONS=OFF")
+
+# Configure release
+if (NOT VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "release")
```
Comment (BillyONeal) [inline ports/bde/portfile.cmake:48]: Is this doing the right thing? Note that it needs to build both debug and release...

File: ports/bde/portfile.cmake [hunk 7]
```diff
@@ -1,133 +1,72 @@
...
-            --build-type=debug
-    WORKING_DIRECTORY ${SOURCE_PATH}
-    LOGNAME configure-${TARGET_TRIPLET}--dbg
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
+    WINDOWS_USE_MSBUILD
+    OPTIONS 
+    ${BDE_OPTIONS}
+    -DBBS_BUILD_SYSTEM=1
+    -DBdeBuildSystem_DIR:PATH=${TOOLS_PATH}/BdeBuildSystem
```
Comment (BillyONeal) [inline ports/bde/portfile.cmake:60]:
```suggestion
    "-DBdeBuildSystem_DIR:PATH=${TOOLS_PATH}/BdeBuildSystem"
```

File: ports/bde/portfile.cmake [hunk 8]
```diff
@@ -1,133 +1,72 @@
...
-message(STATUS "Configuring ${TARGET_TRIPLET}-dbg")
-vcpkg_execute_required_process(
-    COMMAND ${WAF_COMMAND} configure ${CONFIGURE_COMMON_ARGS}
-            --prefix=${CURRENT_PACKAGES_DIR}/debug --out=${SOURCE_PATH_DEBUG}
-            --build-type=debug
-    WORKING_DIRECTORY ${SOURCE_PATH}
-    LOGNAME configure-${TARGET_TRIPLET}--dbg
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
+    WINDOWS_USE_MSBUILD
```
Comment (BillyONeal) [inline ports/bde/portfile.cmake:56]: Stupid question: Why does windows need to be msbuild?

**Comment by BillyONeal** [comment]
Thank you so much! Sorry for the nitpicks...

File: ports/bde/portfile.cmake (hunk 6 above)
Comment (dg0yt) [reply ports/bde/portfile.cmake:48]: What is needed is to add this stuff to `vcpkg_cmake_configure` with
~~~
    OPTIONS_RELEASE
        -DBDE_BUILD_TARGET_OPT=1
    OPTIONS_DEBUG
        -DBDE_BUILD_TARGET_DBG=1
~~~

File: ports/bde/portfile.cmake [hunk 9]
```diff
@@ -1,133 +1,72 @@
...
-if(VCPKG_TARGET_ARCHITECTURE STREQUAL "x86")
-    set(CONFIGURE_COMMON_ARGS ${CONFIGURE_COMMON_ARGS} --abi-bits=32)
-elseif(VCPKG_TARGET_ARCHITECTURE STREQUAL "x64")
-    set(CONFIGURE_COMMON_ARGS ${CONFIGURE_COMMON_ARGS} --abi-bits=64)
-else()
-    message(FATAL_ERROR "Unsupported target architecture: ${VCPKG_TARGET_ARCHITECTURE}")
-endif()
-if(VCPKG_TARGET_IS_WINDOWS)
-    set(CONFIGURE_COMMON_ARGS ${CONFIGURE_COMMON_ARGS} --msvc-runtime-type=static)
+set(BDE_OPTIONS "-DBDE_BUILD_TARGET_CPP17=ON" "-DCMAKE_CXX_STANDARD=17" "-DCMAKE_CXX_STANDARD_REQUIRED=ON" "-DCMAKE_CXX_EXTENSIONS=OFF")
```
Comment (dg0yt) [inline ports/bde/portfile.cmake:45]: I would suggest to get of `BDE_OPTIONS` and add this arguments directly to `vcpkg_cmake_configure` with
~~~
    OPTIONS
       -DBDE_BUILD_TARGET_CPP17=ON
       ...
~~~

File: ports/bde/portfile.cmake [hunk 10]
```diff
@@ -1,133 +1,72 @@
...
+get_filename_component(PYTHON3_EXE_PATH ${PYTHON3} DIRECTORY)
 
 # Acquire BDE Tools and add them to PATH
 vcpkg_from_github(
     OUT_SOURCE_PATH TOOLS_PATH
     REPO "bloomberg/bde-tools"
-    REF d4b1a7670829228f4ec81ecdccc598ce03ae8e80
-    SHA512 80af734c080adb225d5369157301ae0af18e02b1912351e34d23f5f2ba4e19f9ae2b5a367923f036330c9f9afd11a90cdf12680eb3e59b4297a312a1b713f17f
-    HEAD_REF master
+    REF 3.117.0.0
```
Comment (Cheney-W) [inline ports/bde/portfile.cmake:18]:
```suggestion
    REF "${VERSION}"
```

File: ports/bde/portfile.cmake [hunk 11]
```diff
@@ -1,133 +1,72 @@
...
 endif()
 
 # Acquire BDE sources
 vcpkg_from_github(
     OUT_SOURCE_PATH SOURCE_PATH
     REPO "bloomberg/bde"
-    REF 3720d132d0879f19b9084cca62ebc75f1f24e1a3
-    SHA512 234ebb71997f5b7d3951584235ead10f977689cef323ae1c198629a6b1995b1481d8a1515d827c46df10209bdc66e1f3cc7780dafee9ca0ff4172be47c460d78
-    HEAD_REF master
+    REF 3.117.0.0
```
Comment (Cheney-W) [inline ports/bde/portfile.cmake:36]:
```suggestion
    REF "${VERSION}"
```

File: ports/bde/portfile.cmake [hunk 12]
```diff
@@ -1,133 +1,72 @@
...
+    SHA512 810b4a06a08739dcd990751dd543aa7dc58355f9d64a7c96ef0cf45c81501946434db42ad5bcf5d16110d5a463586b587ce09a446136e824298f39a8a871b490
+    HEAD_REF 3.117.0.0
```
Comment (Cheney-W) [inline ports/bde/portfile.cmake:38]:
```suggestion
    HEAD_REF main
```

**Comment by Cheney-W** [comment]
Moving the PR to draft. Please set it to "ready for review" once you respond.

File: ports/bde/portfile.cmake (hunk 3 above)
Comment (adamncasey) [reply ports/bde/portfile.cmake]: I'll leave this variable here for now and just set it using VERSION as you suggest. Perhaps in future these versions will diverge?

**Comment by adamncasey** [comment]
It looks like the CI might take ten hours, I'll mark this ready for review with the idea that my local builds work with these changes.

**Review by BillyONeal — APPROVED** [review]

**Comment by Cheney-W** [comment]
We are fixing the failure of `dcmtk:x64-windows` and `dcmtk:x64-windows-static-md`.

**Comment by Cheney-W** [comment]
Waiting for https://github.com/microsoft/vcpkg/pull/32049 to be merged.

**Comment by Cheney-W** [comment]
Moving the PR to draft. I will set it to "ready for review" once the CI issue fixed.

**Comment by jiayuehua** [comment]
Also should provide the usage file in port bde

**Comment by dg0yt** [comment]
> Is this something known?

Yes. The fix was just merged: #32257.

**Comment by adamncasey** [comment]
```
REGRESSION: halide:x86-windows failed with BUILD_FAILED. If expected, add halide:x86-windows=fail to C:\a\1\s\scripts\azure-pipelines/../ci.baseline.txt.
```

**Comment by dg0yt** [comment]
> ```
> REGRESSION: halide:x86-windows failed with BUILD_FAILED. If expected, add halide:x86-windows=fail to C:\a\1\s\scripts\azure-pipelines/../ci.baseline.txt.
> ```

Baseline issue. Cf. https://github.com/microsoft/vcpkg/pull/32259#issuecomment-1610830059

**Comment by adamncasey** [comment]
The referenced issue is resolved, hopefully this passes now too?

**Comment by FrankXie05** [comment]
@adamncasey Please merge into master branch. :)

**Review by FrankXie05 — APPROVED** [review]

**Comment by BillyONeal** [comment]
Unfortunately, it looks like bde made changes to these so it isn't straightforward to devendor them: https://github.com/bloomberg/bde/commit/5e057e052c28fa9c7c575c06f835e454879bad37 "with Bloomberg extensions"

In particular bde is requiring that a pkgconfig file exists for ryu and it does not upstream. We could add that but without an understanding of what the other changes are I'm not comfortable messing with this on my own.

@adamncasey @cppguru Unfortunately I believe this forces us to remove bde from the curated registry as it can't be installed with other supported ports, and won't be link compatible with anything using the real ryu or pcre2.

**Comment by BillyONeal** [comment]
I filed https://github.com/bloomberg/bde/issues/295

**Comment by cppguru** [comment]
Oleg and Bloomberg is working on this.  I am not, and I'm on leave at the moment. Could you please make an issue in the bde repo or email the team.

**Pull request merged**